### ISSUE TITLE
train test model difference

### ISSUE THREAD

**Issue by fastalgo** [intel/caffe#121]
I finished the training by Caffe run with solver.prototxt and train_val.prototxt files.
Then I do the the testing by Caffe run with train_val.prototxt and .caffemodel files.
But I got the following error. Why the models do not match. I used the same train_val.prototxt:

**F0922 16:55:15.574774 102071 net.cpp:1284] Check failed: target_blobs.size() == source_layer.blobs_size() (3 vs. 5) Incompatible number of blobs for layer layer_64_2_bn1**

**Comment by jgong5** [comment]
May I know what engine you used to run this BatchNorm layer? This issue has been fixed internally and will be upstreamed in the next release soon. But it won't occur if you use MKLDNN engine for all the layers.

**Comment by fastalgo** [comment]
mpirun -np 128 caffe train -engine "MKL2017" --solver=solver.prototxt
caffe test -model test.prototxt -weights log_iter_10000.caffemodel -iterations 5000

**Is there anything wrong?**

**Comment by jgong5** [comment]
Need your model to understand why. But you may try the following patch first.

diff --git a/src/caffe/net.cpp b/src/caffe/net.cpp
index 6edaeb4..cb1b791 100644
--- a/src/caffe/net.cpp
+++ b/src/caffe/net.cpp
@@ -553,18 +553,18 @@ void Net<Dtype>::CompilationRuleOne(const NetParameter& param,

         // If current layer is BatchNorm of MKL2017 engine..
     if (((layer_param->type().compare("BatchNorm") == 0) &&
-       ((layer_param->batch_norm_param().engine() ==
-         BatchNormParameter_Engine_MKL2017)
-       || ((layer_param->batch_norm_param().engine() ==
-           BatchNormParameter_Engine_DEFAULT) &&
-            param.engine().compare("MKL2017") == 0))) ||
+         ((layer_param->batch_norm_param().engine() == BatchNormParameter_Engine_MKL2017) ||
+          ((layer_param->batch_norm_param().engine() == BatchNormParameter_Engine_DEFAULT) &&
+           (layer_param->has_engine() == false)  &&
+           (param.engine().compare("MKL2017") == 0)) ||
+          (param.engine() == "" && layer_param->engine().compare("MKL2017") == 0))) ||
         // If current layer is BatchNorm of MKLDNN engine..
         ((layer_param->type().compare("BatchNorm") == 0) &&
-         ((layer_param->batch_norm_param().engine() == BatchNormParameter_Engine_MKLDNN)
-          || (((layer_param->batch_norm_param().engine() == BatchNormParameter_Engine_DEFAULT) &&
-               (param.engine().compare(0, 6, "MKLDNN") == 0)) ||
-              (param.engine() == "" &&
-               layer_param->engine().compare(0, 6, "MKLDNN") == 0))))) {
+         ((layer_param->batch_norm_param().engine() == BatchNormParameter_Engine_MKLDNN) ||
+          ((layer_param->batch_norm_param().engine() == BatchNormParameter_Engine_DEFAULT) &&
+           (layer_param->has_engine() == false)  &&
+           (param.engine().compare("MKLDNN") == 0)) ||
+          (param.engine() == "" && layer_param->engine().compare("MKLDNN") == 0)))) {
       std::vector<const LayerParameter*> consumer_layer_params;
       GetBlobConsumers(consumer_layer_params,
                        layer_param->top(0),

**Comment by fastalgo** [comment]
You mean I only need to modify net.cpp file? Keep everything else the same?

**Comment by jgong5** [comment]
That's correct. I do not know if that would help without your model file though.

**Comment by fastalgo** [comment]
I can't share my model now, sorry :-) But I tried Intel's ResNet50 model and got the same problem. Can we use the following model as an example? What change do we need to make in this model?

https://github.com/intel/caffe/tree/master/models/intel_optimized_models/resnet_50

**Comment by etaf** [comment]
@fastalgo did you use the same engine for train and test ?

**Comment by fastalgo** [comment]
@etaf Yes, I believe I used the same one.

**Comment by etaf** [comment]
@fastalgo  can I reproduce this issue by train and test the model(https://github.com/intel/caffe/tree/master/models/intel_optimized_models/resnet_50) using MKL2017 engine?

**Comment by fastalgo** [comment]
I think so, please have a try :-)

**Comment by etaf** [comment]
@fastalgo I just tried it, there is no problem when I train and test the model using the same engine.

**Comment by fastalgo** [comment]
@etaf Thanks, I found I did not add -engine "MKL2017" in the testing. So sorry about this! It works for me now.

**Comment by etaf** [comment]
@fastalgo Good to see that, but we still can't reproduce the previous issue, I'll be appreciate if you can construct a prototxt that can reproduce you previous failure. 

F0922 16:55:15.574774 102071 net.cpp:1284] Check failed: target_blobs.size() == source_layer.blobs_size() (3 vs. 5) Incompatible number of blobs for layer **layer_64_2_bn1**

The layer **layer_64_2_bn1** is not the first BatchNorm layer, It should have different engine with previous BatchNorm layer.

**Comment by fastalgo** [comment]
If I remove -engine "MKL2017" in caffe test, I will get this problem.
Thanks,
Yang

**Comment by fastalgo** [comment]
If I remove -engine "MKL2017" in caffe test, I will get this problem.
Thanks

**Comment by etaf** [comment]
@fastalgo  but I can not reproduce even If I remove -engine "MKL2017" in caffe test.

**Comment by fastalgo** [comment]
@etaf Sorry, I can't share my model, which is developed by me and others. Thanks so much!

**Comment by hshen14** [comment]
Close it as we cannot reproduce the issue. Please reopen it when the issue was exposed again.
